File: torture-tests/language-coverage/cross_file_taint_tests.py
"""
=============================================================================
CROSS-FILE TAINT TRACKING TEST SUITE
=============================================================================

PURPOSE: Test taint propagation across file, module, and service boundaries.
These tests verify that Code Scalpel can track taint flow through:

1. Module imports (Python packages)
2. Cross-file function calls
3. Shared data structures
4. Event-driven architectures
5. Message passing systems
6. Database intermediaries

=============================================================================
"""

# =============================================================================
# FILE 1 OF CHAIN: data_source.py
# This file contains the taint SOURCES - user input entry points
# =============================================================================

from typing import Any, Dict, Optional
import json


class UserInputSource:
    """
    TAINT SOURCE: All methods return user-controlled data.
    These represent entry points where tainted data enters the system.
    """

    def get_query_param(self, param_name: str) -> str:
        """TAINT SOURCE: HTTP query parameter."""
        # Simulates: request.args.get(param_name)
        return f"tainted_query_{param_name}"

    def get_form_data(self, field_name: str) -> str:
        """TAINT SOURCE: HTTP form data."""
        # Simulates: request.form.get(field_name)
        return f"tainted_form_{field_name}"

    def get_json_body(self) -> Dict[str, Any]:
        """TAINT SOURCE: JSON request body."""
        # Simulates: request.get_json()
        return {"user_id": "tainted_user_id", "action": "tainted_action"}

    def get_header(self, header_name: str) -> str:
        """TAINT SOURCE: HTTP header."""
        # Simulates: request.headers.get(header_name)
        return f"tainted_header_{header_name}"

    def get_cookie(self, cookie_name: str) -> str:
        """TAINT SOURCE: Cookie value."""
        # Simulates: request.cookies.get(cookie_name)
        return f"tainted_cookie_{cookie_name}"

    def get_path_param(self, param_name: str) -> str:
        """TAINT SOURCE: URL path parameter."""
        # Simulates: request.view_args.get(param_name)
        return f"tainted_path_{param_name}"


class ExternalDataSource:
    """
    TAINT SOURCE: External data that should be treated as untrusted.
    """

    def read_from_queue(self, queue_name: str) -> str:
        """TAINT SOURCE: Message queue data."""
        return "tainted_queue_message"

    def read_from_database(self, query: str) -> Dict[str, Any]:
        """TAINT SOURCE: Database content (stored XSS, etc.)."""
        return {"content": "tainted_db_content"}

    def read_from_file(self, path: str) -> str:
        """TAINT SOURCE: File content (if attacker-controllable)."""
        return "tainted_file_content"

    def call_external_api(self, url: str) -> Dict[str, Any]:
        """TAINT SOURCE: External API response."""
        return {"data": "tainted_api_response"}


# =============================================================================
# FILE 2 OF CHAIN: data_processor.py
# This file processes tainted data - taint must PROPAGATE through here
# =============================================================================

class DataProcessor:
    """
    TAINT PROPAGATION: All methods receive tainted input and return tainted output.
    Taint must be tracked through all transformations.
    """

    def validate_input(self, data: str) -> str:
        """
        TAINT PRESERVING: Validation does NOT sanitize.
        Input and output are both tainted.
        """
        if not data:
            raise ValueError("Empty input")
        if len(data) > 10000:
            raise ValueError("Input too long")
        # IMPORTANT: This does NOT remove taint - it's just validation
        return data

    def transform_input(self, data: str) -> str:
        """
        TAINT PRESERVING: Transformation does NOT sanitize.
        """
        # IMPORTANT: Case change does NOT remove taint
        return data.upper().strip()

    def format_output(self, data: str) -> str:
        """
        TAINT PRESERVING: Formatting does NOT sanitize.
        """
        # IMPORTANT: Formatting does NOT remove taint
        return f"[PROCESSED] {data}"

    def parse_json(self, json_str: str) -> Dict[str, Any]:
        """
        TAINT PRESERVING: Parsed JSON values are tainted.
        """
        # IMPORTANT: JSON parsing does NOT remove taint
        return json.loads(json_str)

    def extract_field(self, data: Dict[str, Any], field: str) -> Any:
        """
        TAINT PRESERVING: Extracted fields are tainted.
        """
        return data.get(field)


class DataTransformer:
    """
    TAINT PROPAGATION through multiple transformation steps.
    """

    def __init__(self, processor: DataProcessor):
        self.processor = processor

    def full_pipeline(self, raw_input: str) -> str:
        """
        TAINT CHAIN: raw_input -> validate -> transform -> format
        All intermediate values are tainted.
        """
        validated = self.processor.validate_input(raw_input)
        transformed = self.processor.transform_input(validated)
        formatted = self.processor.format_output(transformed)
        return formatted  # Still tainted!


# =============================================================================
# FILE 3 OF CHAIN: data_sink.py
# This file contains the taint SINKS - dangerous operations
# =============================================================================

import sqlite3
import subprocess
import os
from typing import List


class DatabaseSink:
    """
    TAINT SINK: SQL operations with tainted data = SQL Injection.
    """

    def __init__(self, db_path: str = ":memory:"):
        self.conn = sqlite3.connect(db_path)

    def execute_query(self, where_clause: str) -> List:
        """
        VULNERABILITY: SQL Injection if where_clause is tainted.
        """
        # TAINT SINK: where_clause reaches SQL query
        query = f"SELECT * FROM users WHERE {where_clause}"
        return self.conn.execute(query).fetchall()

    def insert_data(self, table: str, column: str, value: str) -> None:
        """
        VULNERABILITY: SQL Injection if any parameter is tainted.
        """
        # TAINT SINK: table, column, value all reach SQL
        query = f"INSERT INTO {table} ({column}) VALUES ('{value}')"
        self.conn.execute(query)


class CommandSink:
    """
    TAINT SINK: Command execution with tainted data = Command Injection.
    """

    def execute_command(self, user_arg: str) -> str:
        """
        VULNERABILITY: Command Injection if user_arg is tainted.
        """
        # TAINT SINK: user_arg reaches shell command
        result = subprocess.run(
            f"echo {user_arg}",
            shell=True,
            capture_output=True,
            text=True
        )
        return result.stdout

    def run_with_arg(self, filename: str) -> int:
        """
        VULNERABILITY: Command Injection if filename is tainted.
        """
        # TAINT SINK: filename reaches os.system
        return os.system(f"cat {filename}")


class FileSink:
    """
    TAINT SINK: File operations with tainted paths = Path Traversal.
    """

    def read_file(self, path: str) -> str:
        """
        VULNERABILITY: Path Traversal if path is tainted.
        """
        # TAINT SINK: path controls file access
        with open(path, 'r') as f:
            return f.read()

    def write_file(self, path: str, content: str) -> None:
        """
        VULNERABILITY: Path Traversal if path is tainted.
        """
        # TAINT SINK: path controls file write location
        with open(path, 'w') as f:
            f.write(content)


class EvalSink:
    """
    TAINT SINK: Code execution with tainted data = Code Injection.
    """

    def evaluate(self, expression: str) -> any:
        """
        VULNERABILITY: Code Injection if expression is tainted.
        """
        # TAINT SINK: expression is executed as Python code
        return eval(expression)

    def execute(self, code: str) -> None:
        """
        VULNERABILITY: Code Injection if code is tainted.
        """
        # TAINT SINK: code is executed as Python code
        exec(code)


# =============================================================================
# FILE 4 OF CHAIN: integration.py
# This file connects sources -> processors -> sinks
# This is where cross-file taint must be tracked end-to-end
# =============================================================================

class VulnerableApplication:
    """
    CROSS-FILE TAINT TRACKING TEST

    This class integrates all components and demonstrates
    end-to-end taint flow that Code Scalpel must detect.
    """

    def __init__(self):
        self.source = UserInputSource()
        self.processor = DataProcessor()
        self.transformer = DataTransformer(self.processor)
        self.db_sink = DatabaseSink()
        self.cmd_sink = CommandSink()
        self.file_sink = FileSink()
        self.eval_sink = EvalSink()

    def vulnerable_sql_endpoint(self, request_param: str) -> List:
        """
        VULNERABILITY: Cross-file SQL Injection

        Taint flow:
        1. request_param (TAINTED) - from HTTP request
        2. -> processor.validate_input (still tainted)
        3. -> processor.transform_input (still tainted)
        4. -> db_sink.execute_query (SINK - SQL Injection!)
        """
        user_input = self.source.get_query_param(request_param)
        validated = self.processor.validate_input(user_input)
        transformed = self.processor.transform_input(validated)
        return self.db_sink.execute_query(transformed)

    def vulnerable_command_endpoint(self, filename_param: str) -> str:
        """
        VULNERABILITY: Cross-file Command Injection

        Taint flow:
        1. filename_param (TAINTED) - from HTTP request
        2. -> processor.format_output (still tainted)
        3. -> cmd_sink.execute_command (SINK - Command Injection!)
        """
        user_input = self.source.get_form_data(filename_param)
        formatted = self.processor.format_output(user_input)
        return self.cmd_sink.execute_command(formatted)

    def vulnerable_file_endpoint(self, path_param: str) -> str:
        """
        VULNERABILITY: Cross-file Path Traversal

        Taint flow:
        1. path_param (TAINTED) - from URL path
        2. -> file_sink.read_file (SINK - Path Traversal!)
        """
        user_path = self.source.get_path_param(path_param)
        return self.file_sink.read_file(user_path)

    def vulnerable_eval_endpoint(self) -> any:
        """
        VULNERABILITY: Cross-file Code Injection

        Taint flow:
        1. JSON body (TAINTED) - from HTTP request
        2. -> processor.extract_field (still tainted)
        3. -> eval_sink.evaluate (SINK - Code Injection!)
        """
        json_body = self.source.get_json_body()
        expression = self.processor.extract_field(json_body, "action")
        return self.eval_sink.evaluate(expression)

    def complex_multi_hop_vulnerability(self) -> List:
        """
        VULNERABILITY: Complex multi-hop cross-file taint

        Taint flow (5 hops across 3 files):
        1. get_query_param (SOURCE in data_source.py)
        2. -> validate_input (data_processor.py)
        3. -> transform_input (data_processor.py)
        4. -> format_output (data_processor.py)
        5. -> execute_query (SINK in data_sink.py)
        """
        # Hop 1: Source
        raw = self.source.get_query_param("search")

        # Hops 2-4: Processing
        processed = self.transformer.full_pipeline(raw)

        # Hop 5: Sink
        return self.db_sink.execute_query(processed)


# =============================================================================
# TEST RUNNER
# =============================================================================

def run_cross_file_taint_tests():
    """
    Run all cross-file taint tracking tests.
    """
    print("Cross-File Taint Tracking Test Suite")
    print("=" * 60)
    print("Test cases:")
    print("  1. Source -> Processor -> SQL Sink (SQL Injection)")
    print("  2. Source -> Processor -> Command Sink (Command Injection)")
    print("  3. Source -> File Sink (Path Traversal)")
    print("  4. Source -> Processor -> Eval Sink (Code Injection)")
    print("  5. Complex multi-hop (5 hops, 3 files)")
    print("=" * 60)
    print("Cross-file taint tracking: REQUIRED")
    print("Taint through transformations: REQUIRED")
    print("=" * 60)


if __name__ == "__main__":
    run_cross_file_taint_tests()

File: torture-tests/language-coverage/javascript/cross_file_taint_tests.js
/**
 * =============================================================================
 * JAVASCRIPT CROSS-FILE TAINT TRACKING TEST SUITE
 * =============================================================================
 *
 * PURPOSE: Test taint propagation across JavaScript files, modules, and
 * service boundaries. JavaScript's dynamic nature and module system create
 * unique challenges for cross-file taint tracking.
 *
 * STRUCTURE: This file simulates multiple modules that would typically be
 * in separate files. Code Scalpel must track taint across these boundaries.
 *
 * =============================================================================
 */

const { exec, execSync, spawn } = require('child_process');
const fs = require('fs');
const path = require('path');
const EventEmitter = require('events');

// =============================================================================
// MODULE 1: DATA SOURCES (Simulates data_source.js)
// =============================================================================

/**
 * UserInputSource - All methods return TAINTED data
 */
class UserInputSource {
    /**
     * TAINT SOURCE: HTTP query parameter
     */
    getQueryParam(paramName) {
        return `tainted_query_${paramName}`;
    }

    /**
     * TAINT SOURCE: HTTP form data
     */
    getFormData(fieldName) {
        return `tainted_form_${fieldName}`;
    }

    /**
     * TAINT SOURCE: JSON request body
     */
    getJsonBody() {
        return {
            userId: 'tainted_user_id',
            action: 'tainted_action',
            query: 'tainted_query'
        };
    }

    /**
     * TAINT SOURCE: HTTP header
     */
    getHeader(headerName) {
        return `tainted_header_${headerName}`;
    }

    /**
     * TAINT SOURCE: Cookie value
     */
    getCookie(cookieName) {
        return `tainted_cookie_${cookieName}`;
    }

    /**
     * TAINT SOURCE: URL path parameter
     */
    getPathParam(paramName) {
        return `tainted_path_${paramName}`;
    }

    /**
     * TAINT SOURCE: WebSocket message
     */
    getWebSocketMessage() {
        return { type: 'message', data: 'tainted_ws_data' };
    }
}

/**
 * ExternalDataSource - External data that should be treated as untrusted
 */
class ExternalDataSource {
    /**
     * TAINT SOURCE: Message queue data
     */
    readFromQueue(queueName) {
        return 'tainted_queue_message';
    }

    /**
     * TAINT SOURCE: Database content (stored XSS potential)
     */
    readFromDatabase(query) {
        return { content: 'tainted_db_content' };
    }

    /**
     * TAINT SOURCE: External API response
     */
    async callExternalApi(url) {
        return { data: 'tainted_api_response' };
    }

    /**
     * TAINT SOURCE: File content (if attacker-controllable path)
     */
    readFromFile(filepath) {
        return 'tainted_file_content';
    }
}

// =============================================================================
// MODULE 2: DATA PROCESSORS (Simulates data_processor.js)
// =============================================================================

/**
 * DataProcessor - Transforms data but does NOT sanitize
 */
class DataProcessor {
    /**
     * TAINT PRESERVING: Validation does NOT remove taint
     */
    validateInput(data) {
        if (!data) {
            throw new Error('Empty input');
        }
        if (data.length > 10000) {
            throw new Error('Input too long');
        }
        return data; // STILL TAINTED
    }

    /**
     * TAINT PRESERVING: Transformation does NOT sanitize
     */
    transformInput(data) {
        return data.toUpperCase().trim(); // STILL TAINTED
    }

    /**
     * TAINT PRESERVING: Formatting does NOT sanitize
     */
    formatOutput(data) {
        return `[PROCESSED] ${data}`; // STILL TAINTED
    }

    /**
     * TAINT PRESERVING: JSON parsing preserves taint
     */
    parseJson(jsonStr) {
        return JSON.parse(jsonStr); // All values TAINTED
    }

    /**
     * TAINT PRESERVING: Object spread preserves taint
     */
    mergeWithDefaults(data) {
        const defaults = { safe: true, role: 'user' };
        return { ...defaults, ...data }; // Taint from data spreads
    }

    /**
     * TAINT PRESERVING: Array operations preserve taint
     */
    processArray(items) {
        return items
            .filter(item => item.length > 0)
            .map(item => item.toLowerCase());
        // All items STILL TAINTED
    }
}

/**
 * DataTransformer - Multi-step pipeline that preserves taint
 */
class DataTransformer {
    constructor(processor) {
        this.processor = processor;
    }

    /**
     * TAINT CHAIN: Full pipeline preserves taint through all steps
     */
    fullPipeline(rawInput) {
        const validated = this.processor.validateInput(rawInput);
        const transformed = this.processor.transformInput(validated);
        const formatted = this.processor.formatOutput(transformed);
        return formatted; // STILL TAINTED after 3 transformations
    }

    /**
     * TAINT CHAIN: Async pipeline
     */
    async asyncPipeline(rawInput) {
        const validated = await Promise.resolve(
            this.processor.validateInput(rawInput)
        );
        const transformed = await new Promise((resolve) =>
            setTimeout(() => resolve(this.processor.transformInput(validated)), 10)
        );
        return transformed; // STILL TAINTED after async transformations
    }
}

// =============================================================================
// MODULE 3: DATA SINKS (Simulates data_sink.js)
// =============================================================================

/**
 * DatabaseSink - SQL operations with tainted data = SQL Injection
 */
class DatabaseSink {
    constructor() {
        this.queries = [];
    }

    /**
     * TAINT SINK: SQL Injection if whereClause is tainted
     */
    executeQuery(whereClause) {
        const query = `SELECT * FROM users WHERE ${whereClause}`;
        this.queries.push(query);
        console.log(query); // VULNERABILITY: SQL Injection
        return [];
    }

    /**
     * TAINT SINK: SQL Injection via INSERT
     */
    insertData(table, column, value) {
        const query = `INSERT INTO ${table} (${column}) VALUES ('${value}')`;
        console.log(query); // VULNERABILITY: SQL Injection
    }

    /**
     * TAINT SINK: SQL Injection via dynamic table name
     */
    queryTable(tableName, conditions) {
        const query = `SELECT * FROM ${tableName} WHERE ${conditions}`;
        console.log(query); // VULNERABILITY: SQL Injection
    }
}

/**
 * CommandSink - Command execution with tainted data = Command Injection
 */
class CommandSink {
    /**
     * TAINT SINK: Command Injection via exec
     */
    executeCommand(userArg) {
        exec(`echo ${userArg}`, (error, stdout) => {
            console.log(stdout);
        }); // VULNERABILITY: Command Injection
    }

    /**
     * TAINT SINK: Command Injection via execSync
     */
    executeCommandSync(userArg) {
        return execSync(`cat ${userArg}`).toString(); // VULNERABILITY
    }

    /**
     * TAINT SINK: Command Injection via spawn with shell
     */
    spawnWithShell(command) {
        spawn(command, { shell: true }); // VULNERABILITY
    }
}

/**
 * FileSink - File operations with tainted paths = Path Traversal
 */
class FileSink {
    /**
     * TAINT SINK: Path Traversal via readFile
     */
    readFile(filepath) {
        return fs.readFileSync(filepath, 'utf-8'); // VULNERABILITY
    }

    /**
     * TAINT SINK: Path Traversal via writeFile
     */
    writeFile(filepath, content) {
        fs.writeFileSync(filepath, content); // VULNERABILITY
    }

    /**
     * TAINT SINK: Path Traversal via path.join
     */
    readWithJoin(basePath, userPath) {
        const fullPath = path.join(basePath, userPath);
        return fs.readFileSync(fullPath, 'utf-8'); // VULNERABILITY
    }
}

/**
 * EvalSink - Code execution with tainted data = Code Injection
 */
class EvalSink {
    /**
     * TAINT SINK: Code Injection via eval
     */
    evaluate(expression) {
        return eval(expression); // VULNERABILITY
    }

    /**
     * TAINT SINK: Code Injection via Function constructor
     */
    createFunction(body) {
        return new Function(body)(); // VULNERABILITY
    }

    /**
     * TAINT SINK: Code Injection via setTimeout with string
     */
    delayedEval(code) {
        setTimeout(code, 100); // VULNERABILITY
    }
}

// =============================================================================
// MODULE 4: INTEGRATION (Simulates integration.js)
// =============================================================================

/**
 * VulnerableApplication - Integrates sources, processors, and sinks
 * This demonstrates end-to-end taint flow across module boundaries
 */
class VulnerableApplication {
    constructor() {
        this.source = new UserInputSource();
        this.externalSource = new ExternalDataSource();
        this.processor = new DataProcessor();
        this.transformer = new DataTransformer(this.processor);
        this.dbSink = new DatabaseSink();
        this.cmdSink = new CommandSink();
        this.fileSink = new FileSink();
        this.evalSink = new EvalSink();
    }

    /**
     * CROSS-FILE VULNERABILITY: SQL Injection through processing chain
     */
    vulnerableSqlEndpoint(requestParam) {
        // Source (data_source.js)
        const userInput = this.source.getQueryParam(requestParam);

        // Processing (data_processor.js)
        const validated = this.processor.validateInput(userInput);
        const transformed = this.processor.transformInput(validated);

        // Sink (data_sink.js)
        return this.dbSink.executeQuery(transformed); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Command Injection through formatting
     */
    vulnerableCommandEndpoint(filenameParam) {
        const userInput = this.source.getFormData(filenameParam);
        const formatted = this.processor.formatOutput(userInput);
        return this.cmdSink.executeCommand(formatted); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Path Traversal via path parameter
     */
    vulnerableFileEndpoint(pathParam) {
        const userPath = this.source.getPathParam(pathParam);
        return this.fileSink.readFile(userPath); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Code Injection via JSON body
     */
    vulnerableEvalEndpoint() {
        const jsonBody = this.source.getJsonBody();
        const expression = jsonBody.action;
        return this.evalSink.evaluate(expression); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Multi-hop with async pipeline
     */
    async complexAsyncVulnerability() {
        // Hop 1: Source
        const raw = this.source.getQueryParam('search');

        // Hops 2-4: Async processing
        const processed = await this.transformer.asyncPipeline(raw);

        // Hop 5: Sink
        return this.dbSink.executeQuery(processed); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: External API -> SQL
     */
    async externalApiToSql(apiUrl) {
        const response = await this.externalSource.callExternalApi(apiUrl);
        const data = response.data;
        return this.dbSink.insertData('logs', 'message', data); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Queue message -> Command
     */
    queueToCommand(queueName) {
        const message = this.externalSource.readFromQueue(queueName);
        return this.cmdSink.executeCommand(message); // VULNERABILITY
    }
}

// =============================================================================
// MODULE 5: EVENT-DRIVEN CROSS-FILE TAINT
// =============================================================================

/**
 * EventBus - Demonstrates taint through event-driven architecture
 */
class EventBus extends EventEmitter {
    constructor() {
        super();
        this.setupHandlers();
    }

    setupHandlers() {
        // Handler that receives tainted data
        this.on('userInput', (data) => {
            // TAINT SINK: SQL Injection via event
            const query = `SELECT * FROM events WHERE data = '${data}'`;
            console.log(query); // VULNERABILITY
        });

        // Handler with command execution
        this.on('processCommand', (cmd) => {
            // TAINT SINK: Command Injection via event
            exec(cmd); // VULNERABILITY
        });

        // Handler with file operation
        this.on('readFile', (filepath) => {
            // TAINT SINK: Path Traversal via event
            fs.readFileSync(filepath); // VULNERABILITY
        });
    }
}

/**
 * EventDrivenApp - Cross-file taint through events
 */
class EventDrivenApp {
    constructor() {
        this.source = new UserInputSource();
        this.bus = new EventBus();
    }

    /**
     * CROSS-FILE VULNERABILITY: Taint through event emission
     */
    processUserInput(paramName) {
        const input = this.source.getQueryParam(paramName);
        this.bus.emit('userInput', input); // Taint flows to handler
    }

    /**
     * CROSS-FILE VULNERABILITY: Multiple events with tainted data
     */
    processMultipleInputs() {
        const cmd = this.source.getFormData('command');
        const file = this.source.getPathParam('file');

        this.bus.emit('processCommand', cmd); // VULNERABILITY
        this.bus.emit('readFile', file); // VULNERABILITY
    }
}

// =============================================================================
// MODULE 6: CALLBACK AND CLOSURE CROSS-FILE TAINT
// =============================================================================

/**
 * CallbackProcessor - Taint through callbacks
 */
class CallbackProcessor {
    /**
     * TAINT PRESERVING: Callback receives tainted data
     */
    processWithCallback(data, callback) {
        const processed = data.toUpperCase();
        callback(processed); // Tainted data passed to callback
    }

    /**
     * TAINT PRESERVING: Higher-order function with taint
     */
    createProcessor(transformer) {
        return (data) => {
            const transformed = transformer(data);
            return transformed; // Still tainted
        };
    }
}

/**
 * ClosureApp - Cross-file taint through closures
 */
class ClosureApp {
    constructor() {
        this.source = new UserInputSource();
        this.callbackProcessor = new CallbackProcessor();
    }

    /**
     * CROSS-FILE VULNERABILITY: Taint in callback
     */
    processWithTaintedCallback(paramName) {
        const input = this.source.getQueryParam(paramName);

        this.callbackProcessor.processWithCallback(input, (result) => {
            // TAINT SINK: SQL Injection in callback
            const query = `SELECT * FROM data WHERE x = '${result}'`;
            console.log(query); // VULNERABILITY
        });
    }

    /**
     * CROSS-FILE VULNERABILITY: Taint captured in closure
     */
    createTaintedHandler(paramName) {
        const input = this.source.getQueryParam(paramName);

        // Closure captures tainted input
        return () => {
            // TAINT SINK: Command Injection via closure
            exec(input); // VULNERABILITY
        };
    }

    /**
     * CROSS-FILE VULNERABILITY: Higher-order function with taint
     */
    useHigherOrderProcessor() {
        const input = this.source.getFormData('data');
        const processor = this.callbackProcessor.createProcessor(
            (x) => x.toLowerCase()
        );
        const result = processor(input);

        // TAINT SINK: Path Traversal
        fs.readFileSync(result); // VULNERABILITY
    }
}

// =============================================================================
// MODULE 7: PROMISE CHAIN CROSS-FILE TAINT
// =============================================================================

/**
 * AsyncService - Async operations that preserve taint
 */
class AsyncService {
    async fetchData(query) {
        return { data: `result_for_${query}` }; // Tainted if query is tainted
    }

    async processData(data) {
        return data.toUpperCase(); // Still tainted
    }

    async saveData(data) {
        // TAINT SINK: SQL Injection in async method
        const query = `INSERT INTO results VALUES ('${data}')`;
        console.log(query); // VULNERABILITY
    }
}

/**
 * PromiseChainApp - Cross-file taint through promise chains
 */
class PromiseChainApp {
    constructor() {
        this.source = new UserInputSource();
        this.asyncService = new AsyncService();
    }

    /**
     * CROSS-FILE VULNERABILITY: Promise chain maintains taint
     */
    async promiseChainVulnerability(paramName) {
        const input = this.source.getQueryParam(paramName);

        const result = await this.asyncService.fetchData(input)
            .then(res => this.asyncService.processData(res.data));

        // TAINT SINK: Command Injection after promise chain
        exec(result); // VULNERABILITY
    }

    /**
     * CROSS-FILE VULNERABILITY: Promise.all with tainted data
     */
    async promiseAllVulnerability() {
        const inputs = [
            this.source.getQueryParam('a'),
            this.source.getQueryParam('b'),
            this.source.getQueryParam('c')
        ];

        const results = await Promise.all(
            inputs.map(input => this.asyncService.processData(input))
        );

        // TAINT SINK: SQL Injection with all results
        results.forEach(result => {
            const query = `INSERT INTO log VALUES ('${result}')`;
            console.log(query); // VULNERABILITY (3x)
        });
    }
}

// =============================================================================
// TEST RUNNER
// =============================================================================

function runCrossFileTaintTests() {
    console.log('='.repeat(60));
    console.log('JAVASCRIPT CROSS-FILE TAINT TRACKING TEST SUITE');
    console.log('='.repeat(60));
    console.log('');
    console.log('Module Structure:');
    console.log('  1. Data Sources (UserInputSource, ExternalDataSource)');
    console.log('  2. Data Processors (DataProcessor, DataTransformer)');
    console.log('  3. Data Sinks (DatabaseSink, CommandSink, FileSink, EvalSink)');
    console.log('  4. Integration (VulnerableApplication)');
    console.log('  5. Event-Driven (EventBus, EventDrivenApp)');
    console.log('  6. Callbacks/Closures (CallbackProcessor, ClosureApp)');
    console.log('  7. Promise Chains (AsyncService, PromiseChainApp)');
    console.log('');
    console.log('Cross-File Taint Paths: 20+');
    console.log('Expected Vulnerabilities: 35');
    console.log('='.repeat(60));
}

module.exports = {
    // Sources
    UserInputSource,
    ExternalDataSource,
    // Processors
    DataProcessor,
    DataTransformer,
    // Sinks
    DatabaseSink,
    CommandSink,
    FileSink,
    EvalSink,
    // Integration
    VulnerableApplication,
    // Event-driven
    EventBus,
    EventDrivenApp,
    // Callbacks/Closures
    CallbackProcessor,
    ClosureApp,
    // Async
    AsyncService,
    PromiseChainApp,
    // Runner
    runCrossFileTaintTests
};
